### DISCUSSION TITLE
Any appetite for DBC message/signal merger in the spirit of "Beautiful Soup" ?

### DISCUSSION THREAD

**Discussion by mdear** [cantools/cantools#669] [Ideas]
Hi, it's Myles Dear, software director from https://ktechnology.ca working on a heavy duty mobile electric truck project.  

Is there any appetite on the part of the most senior cantools maintainers for me to generate a pull request that provides python code gen and message merge capabilities ?

I figured I should ask this before foisting my pull request on you, I also figured you may have some architectural-level comments that may constrain my pull request.    

I've tasked myself with loading several DBC files from one of our suppliers (for the purposes of automatic Wireshark dissection and creation of python stubs similar to your c language code generator) and, although each DBC file loads fine individually, when loaded together both message and signal level conflicts ensue.  Due to the nature of the implementation of this company's various products, some conflicts cannot be helped (code bases get copied when new products are created and they then evolve apart).  

I've dug in deeper and informed the supplier that they must ensure their DBC files play nice together.  Apparently the tooling they use to edit their DBC files ensure consistency on the level of a particular file but don't detect collisions among multiple DBC files from differing product families within the same company.  All they have committed to is to separate the messages that are sent to a unicast address versus messages that are sent to a multicast address so that`s a step forward, I suppose.

I've managed to design and implement a message and signal merge capability that is mux-aware, and I'm just finishing up an activity to feed those merged messages and signals back into cantools/database/can/message (and signal) in order to reconstruct the object environment necessary to encode/decode against these merged messages.

I've enclosed the two troublesome DBC files for discussion purposes.  I've refactored the vendor DBC files a little but there are collisions at both the message and signal levels that I have been able to resolve by merging without losing too much of the original DBC content.  My merger code will refuse to merge if two different signals overlap outright and share a mux id or if two mux signals share different offsets/lengths.  I'll also generate non-critical warnings that inform the user specific conflicts were found and resolved as part of the merge process.  

Thanks for reading to the end !

[BlinkMarineKeyboxAnd5ButtonKeypad.zip](https://github.com/user-attachments/files/15855958/BlinkMarineKeyboxAnd5ButtonKeypad.zip)

**Comment by geo4711** [comment]
Hi Myles,
although I'm not one of the senior maintainers, I'd like to share some thoughts about this topic.
Some years ago I tried to merge some dbcs, too. However, I realized that there are many, many aspects to deal with if you want to have a "fully working" merge feature. For my special use case, I could omit/ignore most of these aspects. But when providing a general merge feature to "clueless" cantools users, I had decided to not implement this. Some examples of those nasty aspects:
* messages of different dbcs might have same identifier, but different name (and vice versa)
* msgs may have different sender nodes (or different list/sets of sender nodes); even if the nodes might have the same name, their attributes might differ ...)
* different dbcs might have different sets of attributs, (including different enum values/value ranges, different default values, ...)
* * applies seperately to the dbc itself, to messages, signals and as far as I remember also to nodes and environment variables
* * applies even to those objects that are not intended to be merged

* different dbcs might have some settings that make their content incompatible (example: I'm not sure if it is allowed to have J1939 frames and CANFD frames in the same file. At least my Vector DBC++ does not support this) ...
* and of course all the stuff you already mentioned (e. g. multiplex aspects).

For these reasons, in my opinion a "fully working" feature would need to have either of these features:
* interactive mode/GUI to see which data causes conflicts/needs individual decision
* lots of options in order to run automatically
* some low level features (e. g. merge two signal lists)

Maybe you've already found a smart way to deal with some of these aspects.
Hope my concerns wont discourage you ...

BR
Martin

**Reply by mdear** [reply]
And, yes, I'm already dealing with many of the issues you mentioned. 

First, I normalize the canid by treating a broadcast destination 0xff specially but other normalizing a nonzero source of destination byte to 0x01. 

When merging two messages with identical normalized canids I first deal with low hanging fruit : rewriting an original message's multiplexer to accept new nonconflicting code points and I rewrite new candidate message content to point to the original mux signal name and then merge back into the existing message. Some unusual names are expected when Frankensteining this content together (it's alive ...). I did say "best effort", right?

I then identify serious nonconformances, including a candidate message with identical canid but incompatible mux signal that is in a different part of the frame than the original signal. I reject the candidate message for merging in this case. 

And, for a candidate message with conflicting mux ids, I must destructively remove/excise them from the merge result to prevent serious collisions. That's why the order of DBC files presented is important since already loaded messages are prioritized higher than incumbent merge candidate messages. 

All these cases are logged at warning or error levels to the module logger for posterity. 

In this case, necessity appears to be the mother of invention. There's nothing like a deadline to "encourage" creativity !

**Comment by andlaus** [comment]
I agree with @geo4711: It can be done but it will be a lot of work and you have to deal with plenty of corner cases (which you need to detect and raise exceptions or warn). Besides this, I do not really see a huge number of benefits for such a `merge` tool compared to loading the files individually and trying to decode a message using the first database which knows about it, but if you want to give it a shot at implementing it, I'll certainly will consider it for merging...

**Reply by mdear** [reply]
Thanks for the great thoughts. 

So far the implementation is holding for my small use case but the vendor really does have two devices whose messages can conflict on the wire so I prioritize which device I wish to communicate with and added a destructive design which literally removes any conflicting signal's mux id and mux signal pointers while merging any message that is mergable. 

I asked the vendor to migrate the identified inter product conflicts to newer code points that do not collide and add a "legacy mode" in which both legacy and new signals are transmitted. When this legacy feature is disabled the conflicting code points would  no longer be sent. I hope they clean up their products to better interoperate on the wire. 

I think I'll wait to generate a pull request until I kick the tires some more and add a few more DBC files to the mix but for now I think the merge design is working. Just yesterday I implemented a design that allows the merged messages to be recreated in the form of native cantools database/message/signal objects (hence the need for the destructive merge design to ensure internal cantools validation did not fail). I am also considering using cantools' ability to rewrite a DBC file from internal content as this would help to close the loop. 

Given the range of corner cases this merge design will always be "best effort" but I find it useful to "make a silk purse out of a sow's ear" and at least have some degree of automated interactions with conflicting devices (this is the Beautiful Soup philosophy). Life is messy and I hope to be able to contribute this merger that will try to make it at least partially possible to interact with conflicting devices to the largest extent the vendor allows. 

I'll also consider the can-fd angle because I have told the owner of our company that it is likely we'll need to upgrade to it. Semantically, I don't see a reason why the model wouldn`t work, since the DBC would just specify message length that could exceed 8 bytes, from my understanding.  

To your point about the necessity of such a tool, in order to achieve real time efficiency in a production system I suppose we could preload multiple databases and check them individually but in order to achieve the same result more efficiently seems to be a worthwhile goal to me so I'll keep chipping away at the stone block and see if there's a sculpture underneath, metaphorically speaking.  Creating recognized differentiation is an important part of building any business. 

Watch this space !  More soon.

**Comment by mdear** [comment]
Working code may be found here (not a PR yet, more tire kicking needed), but this will help you get the gist of the changes : 

https://gist.github.com/mdear/fe9f411a9376cb6ebb14ab7f630774ca
